Homework #8 - Networking

prometey-dev opened the pull request:
**link to homework** - https://docs.google.com/document/d/e/2PACX-1vTK-zl3Cq6YTZ_1e0HWsJEDEwNHHylJQzbLDs7D3GQ1mwezBIWfxF97mnMPQbAJ5Q/pub

### What was done:

1. Added work with network requests, integration with api www.themoviedb.org. Uploading popular movies to list.
2. Added Token Interceptor
3. Added error handling
4. Added loader
5. Architectire Mvvm refact
6. Added dependency injection with Koin

### What's left to do:

1. Implement pagination to downloading films with Pagination Library from Google
2. Added unit tests

### Results:

**_Movie list loaded_**

![image](https://user-images.githubusercontent.com/6102319/103486682-90c1df00-4e10-11eb-89e0-2de5300175e5.png)

**_Movie list is loading_**

![image](https://user-images.githubusercontent.com/6102319/103486784-63c1fc00-4e11-11eb-8010-2430897fe7f7.png)

**_Movie details loaded_**

![image](https://user-images.githubusercontent.com/6102319/103486818-b4d1f000-4e11-11eb-860b-60ef98d035ce.png)

![image](https://user-images.githubusercontent.com/6102319/103486827-cc10dd80-4e11-11eb-9f31-6ab81f3add27.png)

**_Movie details is loading_**

![image](https://user-images.githubusercontent.com/6102319/103486871-290c9380-4e12-11eb-8f45-90343334d120.png)

prometey-dev committed:
add networking

prometey-dev committed:
add token interceptor

prometey-dev committed:
add handling errors

prometey-dev committed:
add loading handler

prometey-dev committed:
refact architecture mvvm

prometey-dev committed:
add dependency injection with Koin

prometey-dev committed:
add abstract class BaseUseCases and add handle http errors

prometey-dev committed:
refact MoviesRepository

emartynov commented on app/build.gradle:
```diff
@@ -45,9 +47,17 @@ dependencies {
+    implementation 'com.squareup.retrofit2:retrofit:2.9.0'
+    implementation 'com.squareup.retrofit2:converter-moshi:2.6.2'
+    implementation 'com.squareup.moshi:moshi-kotlin:1.11.0'
+    implementation "org.koin:koin-android-viewmodel:2.0.1"
```
Some are double quoted and some single quoted, why?

prometey-dev replied:
I will do the same

emartynov commented on app/src/main/java/ru/prometeydev/movie/di/ViewModelModule.kt:
```diff
@@ -0,0 +1,11 @@
+import ru.prometeydev.movie.ui.moviesdetails.MoviesDetailsViewModel
+import ru.prometeydev.movie.ui.movieslist.MoviesListViewModel
+
+val viewModelModule = module {
```
Minor, that is OKE, but in real project this will grow enormously for every view model

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/BaseUseCases.kt:
```diff
@@ -0,0 +1,51 @@
+
+    private val dispatcher: CoroutineDispatcher = Dispatchers.IO
+
+    private val moshi = Moshi.Builder().build()
```
You can also use Koin to have one moshi instance in app

prometey-dev replied:
ok, I'll do it

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/BaseUseCases.kt:
```diff
@@ -0,0 +1,51 @@
+
+    private fun handleHttpException(httpException: HttpException): Throwable {
+        val errorMessage = httpException.response()?.errorBody()?.string()
+        val jsonAdapter = moshi.adapter(ErrorResponse::class.java)
```
Ah here it is used, you even can inject error adapter

prometey-dev replied:
ok, I'll do it

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/BaseUseCases.kt:
```diff
@@ -0,0 +1,51 @@
+        val errorMessage = httpException.response()?.errorBody()?.string()
+        val jsonAdapter = moshi.adapter(ErrorResponse::class.java)
+        val errorResponse = jsonAdapter.fromJson(errorMessage)
+        return Exception(errorResponse?.message)
```
Why no special exception class?

prometey-dev replied:
i will fix

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/BaseUseCases.kt:
```diff
@@ -0,0 +1,51 @@
+            is ConnectException,
+            is SocketTimeoutException,
+            is SocketException,
+            is ProtocolException -> ConnectionErrorException()
```
I would not create new exception classes but just use existing one. All of these are extending from IOException.

prometey-dev replied:
i will fix

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/domain/ActorDto.kt:
```diff
@@ -0,0 +1,11 @@
+
+@JsonClass(generateAdapter = true)
+data class ActorDto(
+    @Json(name = "id") val id: Int,
```
why to repeat id in the @Json?

prometey-dev replied:
This is my habit. I read somewhere about such a recommendation to annotate each field. I really don’t know how bestpracticles are.

emartynov replied:
Can you provide a link? The moshi will generate the adapter and it will have correct string for parsing so you will not have any problem with obfuscation. As for me it is noise - I can not see which filed on BE are different from my model. 

But up to you. Your style is yours. But you have to have good grounded explanation why you do it.

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/domain/CastsDto.kt:
```diff
@@ -0,0 +1,10 @@
+
+@JsonClass(generateAdapter = true)
+data class CastsDto(
+    @Json(name = "id") val id: Int,
```
Same

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/domain/ConfigurationDto.kt:
```diff
@@ -0,0 +1,9 @@
+
+@JsonClass(generateAdapter = true)
+class ConfigurationDto(
+    @Json(name = "images") val images: ImagesDto
```
Same but about images

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/domain/GenreDto.kt:
```diff
@@ -0,0 +1,10 @@
+
+@JsonClass(generateAdapter = true)
+data class GenreDto(
+    @Json(name = "id") val id: Int,
```
Same about two properties

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/domain/GenresDto.kt:
```diff
@@ -0,0 +1,9 @@
+
+@JsonClass(generateAdapter = true)
+class GenresDto(
+    @Json(name = "genres") val genres: List<GenreDto>
```
Same

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/domain/MovieDto.kt:
```diff
@@ -0,0 +1,17 @@
+import com.squareup.moshi.JsonClass
+
+@JsonClass(generateAdapter = true)
+data class MovieDto(
```
What is difference between movie dto and movie details dto?

prometey-dev replied:
I think you can make a general class. Just a query that returns a list of popular films in an element of one movie stores truncated information, and a query that returns movie details by id contains a more extensive set of data. Therefore, I passed two different classes. But I think you can limit yourself to one class.

emartynov replied:
Up to you, but I would probably use one class

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/network/MoviesApiProvider.kt:
```diff
@@ -0,0 +1,68 @@
+        private const val CONNECTION_TIMEOUT = 15L
+        private const val READ_TIMEOUT = 15L
+
+        private const val API_KEY = "6919274aeda3b6ad96593bb0eafb4afd"
```
Bad practice to have token in code

prometey-dev replied:
And if you put it in build.gradle? Where is the best place to store?

emartynov replied:
The best is to have a local file or environment variable that is inserted on the build phase.

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/network/MoviesApiProvider.kt:
```diff
@@ -0,0 +1,68 @@
+    private val okHttpClient: OkHttpClient
+
+    init {
+        val tokenInterceptor = TokenInterceptor()
```
Mental - try to keep constructors really light weight. This is heavy and can crash with some exception. Hard to test and mind blowing what to do with crash in runtime. Also possible performance penalty.

emartynov commented on app/src/main/java/ru/prometeydev/movie/ui/base/Event.kt:
```diff
@@ -0,0 +1,29 @@
+package ru.prometeydev.movie.ui.base
+
+data class Event<out T>(
```
Check the sealed classes.

sealed class Result<T> {
  class Error(val errorMessage: String?) : Result<Nothing>
 ....
}

emartynov replied:
Then you don't need Status enum at all

emartynov commented on app/src/main/java/ru/prometeydev/movie/ui/base/BaseViewModel.kt:
```diff
@@ -0,0 +1,26 @@
+
+open class BaseViewModel : ViewModel() {
+
+    fun <T> requestWithLiveData(
```
protected/internal?

prometey-dev replied:
ok, protected

emartynov commented on app/src/main/java/ru/prometeydev/movie/ui/moviesdetails/MoviesDetailsViewModel.kt:
```diff
@@ -2,24 +2,23 @@ package ru.prometeydev.movie.ui.moviesdetails
+) : BaseViewModel() {
 
-    private val _mutableMovieState = MutableLiveData<Movie>()
+    private val _mutableLiveData = MutableLiveData<Event<MovieDetails>>()
```
Duplication - you can use generics and move mutable live data and live data in the base model

emartynov reviewed (commented):
Looks solid!

Some comments to think or change

prometey-dev committed:
Add pagination loading movies

prometey-dev committed:
mnor fixes

prometey-dev committed:
put api key into keystore.properties

prometey-dev committed:
add vertival scrollbars into list movies

prometey-dev commented:
1) Corrected some of the comments.
2) Added pagination when scrolling through the list of films with subsequent loading. I used Paging3 and Flow here. Also, when loading errors occur, a Retry button is added to the fotter and a message pops up. You can repeat the download.

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/BaseUseCases.kt:
```diff
@@ -30,22 +28,18 @@ abstract class BaseUseCases {
         val jsonAdapter = moshi.adapter(ErrorResponse::class.java)
         val errorResponse = jsonAdapter.fromJson(errorMessage)
-        return Exception(errorResponse?.message)
+        return Throwable(errorResponse?.message)
```
А чего throwable теперь?

prometey-dev replied:
Need to return some CustomThrowable like ErrorResponseException?

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/MoviesPagingSource.kt:
```diff
@@ -0,0 +1,31 @@
+            val movies = repository.loadMovies(page)
+            LoadResult.Page(
+                data = movies,
+                prevKey = if (page == STARTING_PAGE_INDEX) null else page - 1,
```
Is null marker about no data? Is it library expectations or your implementation?

prometey-dev replied:
This implementation example was borrowed by me from codelab https://developer.android.com/codelabs/android-paging#4
those. if we are on the first page, then the previous key must be null.

prometey-dev replied:
The same principle is applied in this article https://proandroiddev.com/how-to-use-the-paging-3-library-in-android-5d128bb5b1d8

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/MoviesPagingSource.kt:
```diff
@@ -0,0 +1,31 @@
+            LoadResult.Page(
+                data = movies,
+                prevKey = if (page == STARTING_PAGE_INDEX) null else page - 1,
+                nextKey = if (movies.isEmpty()) null else page + 1
```
This is a bit un-efficient - you movie load has number of total pages. So you can just compare that as marker of the end.

prometey-dev replied:
Yes, I agree, I will.

emartynov commented on app/src/main/java/ru/prometeydev/movie/model/MoviesRepository.kt:
```diff
@@ -15,11 +15,11 @@ class MoviesRepository(network: MoviesApiProvider): BaseUseCases() {
+    suspend fun loadMovies(page: Int): List<Movie> = execute {
         baseImageUrl = getImageUrl()
 
         val genresMap = loadGenres()
```
Why to load genres all the time?

prometey-dev replied:
Is it possible to transfer genres loading to the constructor?

emartynov commented on app/src/main/java/ru/prometeydev/movie/ui/base/Result.kt:
```diff
@@ -0,0 +1,8 @@
+
+sealed class Result {
+    object Loading : Result()
+    data class Success<out T>(val data: T?) : Result()
```
Why nullable?

prometey-dev replied:
Ok, nullable remove

emartynov commented on app/src/main/java/ru/prometeydev/movie/ui/base/Result.kt:
```diff
@@ -0,0 +1,8 @@
+sealed class Result {
+    object Loading : Result()
+    data class Success<out T>(val data: T?) : Result()
+    data class Error(val message: String?) : Result()
```
Same, use "Unexpected error" as default

emartynov commented on app/src/main/java/ru/prometeydev/movie/ui/moviesdetails/MoviesDetailsViewModel.kt:
```diff
@@ -5,22 +5,21 @@ import androidx.lifecycle.MutableLiveData
 
 class MoviesDetailsViewModel(
     private val repository: MoviesRepository
 ) : BaseViewModel() {
```
Do you need session with Generics?

prometey-dev replied:
And why are generics needed here?

emartynov commented on app/src/main/java/ru/prometeydev/movie/ui/movieslist/MoviesListFragment.kt:
```diff
@@ -51,8 +82,10 @@ class MoviesListFragment : BaseFragment() {
 
     @Suppress("UNCHECKED_CAST")
     override fun <T> bindViews(data: T) {
```
Why Generics here?

prometey-dev replied:
Because in MovieDetailsFragment in bindView we pass the MovieDetails type, and in MovieListFragment we pass PagingData <Movie>.

emartynov reviewed (commented):
Put more comments

prometey-dev committed:
minor fixes

prometey-dev commented:
Added new fixes

prometey-dev committed:
make BaseViewModel with generic

prometey-dev committed:
remove Suppress UNCHECKED_CAST

emartynov commented on app/src/main/java/ru/prometeydev/movie/ui/base/BaseFragment.kt:
```diff
@@ -0,0 +1,65 @@
+    abstract fun startObserve()
+    abstract fun destroyViews()
+    abstract fun loadData()
+    abstract fun bindViews(data: Any)
```
You can make base fragment also generic to avoid `Any` usage

prometey-dev replied:
I will do this in a separate PR and also use Flow for both models

emartynov reviewed (approved):
There are still things to clean:
1) Abandone live data and use Flow for both models
2) Make base fragment generic to avoid casting data in the bind data

But I think it is already so long that this PR is live to approving it with hope of the better things in the future.
